zend_abstract_interface/functions/functions.h

```diff
@@ -0,0 +1,32 @@
+#ifndef ZAI_FUNCTIONS_H
+#define ZAI_FUNCTIONS_H
+
+#include <main/php.h>
+#include <stdbool.h>
+
+/* Work in progress
+ *
+ * The long-term goal is to provide ZAI data-structure shims so that the
+ * following APIs are the same across all PHP versions. For now we will provide
+ * ZAI seams using native Zend Engine data structures.
+ */
+
+/* Calls a function with 'argc' number of zval arguments. Caller owns the
+ * arguments and the 'retval'. Caller is responsible for freeing any refcounted
+ * arguments. Caller must provide a pointer to a stack-allocated zval for the
+ * return value, 'retval'. In error cases, 'retval' will be set to IS_UNDEF.
+ * Caller must dtor the 'retval' after the call:
+ *
+ *   zval_ptr_dtor(&retval);
+ *
+ * Functions cannot be called outside of a request context so this MUST be
+ * called from within a request context (after RINIT and before RSHUTDOWN).
+ */
+bool zai_call_function(const char *name, size_t name_len, zval *retval, int argc, ...);
```

Comment:
do we want to add like ~6 lines of macro hell to avoid manually passing the arg count? Looks to me like such an easy pitfall to mess up. The resulting macro would look similarly to https://github.com/php/php-src/blob/d60bc0e2d91d299c6cb03cb48672479bcffd03a7/Zend/zend_ast.h#L227.

Reply:
I like this idea. I'll tackle this on Monday. Thanks for the suggestion!

Reply:
Thanks for helping me figure out the compiler error (missing `retval` 🤦). This was addressed in 0648f8d and f4e4cdc.